### ISSUE TITLE
Panic in core/src/heaviest_subtree_fork_choice.rs, thread solReplayStage

### ISSUE THREAD

**Issue by ruuda** [solana-labs/solana#32880]
#### Problem

One of our Solana mainnet nodes panicked. It was not validating at the time.

```
thread 'solReplayStage' panicked at 'a bank at last_voted_slot((212118116, ugZWxPNYKgnkE1h71oek7Xvr9vDvcwRS1PZ8HcgcRGv)) is a frozen bank so must have been added to heaviest_subtree_fork_choice at time of freezing', core/src/heaviest_subtree_fork_choice.rs:874:29
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
   2: <solana_core::heaviest_subtree_fork_choice::HeaviestSubtreeForkChoice as solana_core::fork_choice::ForkChoice>::select_forks
   3: solana_core::replay_stage::ReplayStage::new::{{closure}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2023-08-18T07:20:50.607088657Z ERROR solana_metrics::metrics] datapoint: panic program="validator" thread="solReplayStage" one=1i message="panicked at 'a bank at last_voted_slot((212118116, ugZWxPNYKgnkE1h71oek7Xvr9vDvcwRS1PZ8HcgcRGv)) is a frozen bank so must have been added to heaviest_subtree_fork_choice at time of freezing', core/src/heaviest_subtree_fork_choice.rs:874:29" location="core/src/heaviest_subtree_fork_choice.rs:874:29" version="\"1.14.23 (src:devbuild; feat:1879391783)\""
```

This is running Solana 1.14.23.

[Full validator logs since the earliest occurrence of `ugZWxPNYKgnkE1h71oek7Xvr9vDvcwRS1PZ8HcgcRGv`.](https://github.com/solana-labs/solana/files/12377051/2023-08-18-solana.txt)

The code is this: https://github.com/solana-labs/solana/blob/68c6fd139fb7b2650be5852270cee703cae6acbc/core/src/heaviest_subtree_fork_choice.rs#L866-L878

It was originally introduced in https://github.com/solana-labs/solana/commit/cb8661bd49bb3591360589cba73d3593249b9fcc#diff-f1f89ae82910cdfebacc27d6ef29a9a55dc89eea426a8e4fcb1586f4ca8da55cR501-R513 by @ryoqun, although the conditions around it have changed since then. @ryoqun, do you have any clue what might be going on here?

#### Proposed Solution

I am not familiar enough with the replay stage to investigate what went wrong here. Let me know if there is anything I can do to help diagnose this further.

**Comment by ruuda** [comment]
It happened a second time to the same node:

```
solReplayStage' panicked at 'a bank at last_voted_slot((212139600, EQ1VXBxKTLePTpAwfYTWFusgGRzswiepwZntDrKXqG3R)) is a frozen bank so must have been added to heaviest_subtree_fork_choice at time of freezing', [2023-08-18T09:58:55.224186191Z INFO  solana_metrics::metrics] datapoint: bank_weight slot=212139603i weight="0"
core/src/heaviest_subtree_fork_choice.rs:874:29
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
   2: <solana_core::heaviest_subtree_fork_choice::HeaviestSubtreeForkChoice as solana_core::fork_choice::ForkChoice>::select_forks
   3: solana_core::replay_stage::ReplayStage::new::{{closure}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2023-08-18T09:58:55.240753496Z ERROR solana_metrics::metrics] datapoint: panic program="validator" thread="solReplayStage" one=1i message="panicked at 'a bank at last_voted_slot((212139600, EQ1VXBxKTLePTpAwfYTWFusgGRzswiepwZntDrKXqG3R)) is a frozen bank so must have been added to heaviest_subtree_fork_choice at time of freezing', core/src/heaviest_subtree_fork_choice.rs:874:29" location="core/src/heaviest_subtree_fork_choice.rs:874:29" version="\"1.14.23 (src:devbuild; feat:1879391783)\""
```
[Logs since first occurrence of `EQ1VXBxKTLePTpAwfYTWFusgGRzswiepwZntDrKXqG3R`.](https://github.com/solana-labs/solana/files/12378255/solana-panic-2023-08-18-9-58-sanitized.log)

**Comment by ruuda** [comment]
It happened a third time to the same node:

```
thread 'solReplayStage' panicked at 'a bank at last_voted_slot((212165388, 4he27KybCzcVqhPkcL8ZG4WXTg2ZkspsJk3ZBdHHV9au)) is a frozen bank so must have been added to heaviest_subtree_fork_choice at time of freezing', core/src/heaviest_subtree_fork_choice.rs:874:29
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
   2: <solana_core::heaviest_subtree_fork_choice::HeaviestSubtreeForkChoice as solana_core::fork_choice::ForkChoice>::select_forks
   3: solana_core::replay_stage::ReplayStage::new::{{closure}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2023-08-18T13:07:02.218784854Z ERROR solana_metrics::metrics] datapoint: panic program="validator" thread="solReplayStage" one=1i message="panicked at 'a bank at last_voted_slot((212165388, 4he27KybCzcVqhPkcL8ZG4WXTg2ZkspsJk3ZBdHHV9au)) is a frozen bank so must have been added to heaviest_subtree_fork_choice at time of freezing', core/src/heaviest_subtree_fork_choice.rs:874:29" location="core/src/heaviest_subtree_fork_choice.rs:874:29" version="\"1.14.23 (src:devbuild; feat:1879391783)\""
```

[Logs since slightly before the first occurrence of `4he27KybCzcVqhPkcL8ZG4WXTg2ZkspsJk3ZBdHHV9au`](https://github.com/solana-labs/solana/files/12380156/solana-panic-2023-08-18-13-07-sanitized.log).

We run multiple nodes and it’s happening only on this machine, but so far we have no evidence it’s a problem with the host — no signs of data corruption etc.

**Comment by ruuda** [comment]
The same issue occurred many more times over the weekend, now also on a different machine.

**Comment by ruuda** [comment]
Some people are posting the same panic on Discord ([1](https://discord.com/channels/428295358100013066/689412830075551748/1141997540166860800), [2](https://discord.com/channels/428295358100013066/560174212967432193/1142187134275571822)). [Trent says](https://discord.com/channels/428295358100013066/560174212967432193/1142207749963395152):

> it's going to crash every time the secondary observes the primary voting at the tip

**Comment by ruuda** [comment]
https://github.com/solana-labs/solana/pull/32894 references the same problem.

**Comment by AshwinSekar** [comment]
Fix has been released in 1.14.25 and 1.16.10 .

**Comment by Szymongib** [comment]
It is more rare, but we still see it on 1.14.27:
```
Sep 28 13:32:02 solana-validator[746621]: [2023-09-28T13:32:02.878695407Z ERROR solana_metrics::metrics] datapoint: panic program="validator" thread="solReplayStage" one=1i message="panicked at 'a bank at last_voted_slot((220308112, ArmtDGoACSv4jSwsY1XyqWpj9eAFhUsVM7j6n3EBHRrz)) is a frozen bank so must have been added to heaviest_subtree_fork_choice at time of freezing', core/src/heaviest_subtree_fork_choice.rs:874:29" location="core/src/heaviest_subtree_fork_choice.rs:874:29" version="\"1.14.27 (src:devbuild; feat:3351361749)\""
Sep 28 13:32:02 solana-validator[746621]: note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Sep 28 13:32:02 solana-validator[746621]:    3: solana_core::replay_stage::ReplayStage::new::{{closure}}
Sep 28 13:32:02 solana-validator[746621]:    2: <solana_core::heaviest_subtree_fork_choice::HeaviestSubtreeForkChoice as solana_core::fork_choice::ForkChoice>::select_forks
Sep 28 13:32:02 solana-validator[746621]:              at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
Sep 28 13:32:02 solana-validator[746621]:    1: core::panicking::panic_fmt
Sep 28 13:32:02 solana-validator[746621]:              at ./rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
Sep 28 13:32:02 solana-validator[746621]:    0: rust_begin_unwind
```
